### ISSUE TITLE
param loop over bool range implicitly casts iterator variable to int unexpectedly

### ISSUE THREAD

**Issue by ian-bertolacci** [chapel-lang/chapel#9860]
A user (me) might expect that the following code had no compilation errors.

```chpl
proc foo( param i : bool ) {
  writeln( i );
}

for param i in false..true {
  foo( i );
}
```

However the following error is generated:
```
test.chpl:6: error: unresolved call 'foo(0)'
test.chpl:1: note: candidates are: foo(param i: bool)
```

It appears that the loop iterator `i ` is being implicitly converted to an int, as illustrated by the below code:
```chpl
for param i in false..true {
  writeln( (i.type : string, i) );
}
```

which produces the output:
```
(int(64), 0)
(int(64), 1)
```

```
> chpl --version
chpl version 1.17.1
Copyright (c) 2004-2018, Cray Inc.  (See LICENSE file for more details)
```

**Comment by ian-bertolacci** [comment]
> "but would anyone ever use them?"
The only reason I am using them is because I cant do `for param i in [true, false] do ....` (unless there is a way to do this and I just don't know about it.

Your coercion to int explanation makes the looping behavior make sense, especially for when I tried `for i in false..true do ...` and got an empty loop.

**Comment by bradcray** [comment]
Oh, I overlooked the `param` part...  it will take additional work to support param loops over ranges of bools or enums (but I can see why that would be a worthy feature...).

**Comment by bradcray** [comment]
Probably obvious, but just in case: a workaround for this today would be to put the body of the loop into a function taking a param bool argument and call it with both true and false.  I..e, rather than:

```chapel
for param b in false..true do
    ...body using b...
```

use:

```
body(false);
body(true);
proc body(param b: bool) {
  ... body using b ...
}
```

This is not to say the request isn't a very reasonable one... just trying to keep you unblocked.

**Comment by ian-bertolacci** [comment]
I found that you can just cast:
```chpl
proc foo( param i : bool ) {
  writeln( i );
}

for param i in false..true {
  foo( i : bool );
}
```

The use case for this was testing LayoutCS across all combinations of sorted and whatever else I was throwing at it.

**Comment by bradcray** [comment]
Oh yeah, that's an even better workaround...

**Comment by bradcray** [comment]
Hi Ian -- I'm going to close this in favor of #9866.

**Comment by bradcray** [comment]
Hi Ian -- I've now implemented this (including the param case) in PR #9916.
